### ISSUE TITLE
[node-v6] Got exception parsing to JS on Node v6 

### ISSUE THREAD

**Issue by dgaubert** [CartoDB/carto#24]
For the following css input:

```
#layer[name=~".*Cairo*"] {
  marker-width: 14;
}
```

RendererJS throws:

```
(new carto.RendererJS({ debug: true })).render(css);
---
if((8388607 & (1 << ctx.zoom)) && data['name'] =~ '.*Cairo*'){_value = 14;}
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ReferenceError: Invalid left-hand side in assignment
      at Object._createFn (/vagrant/repos/carto/lib/carto/renderer_js.js:153:12)
      at Object.parse (/vagrant/repos/carto/lib/carto/renderer_js.js:245:40)
      at Object.setStyle (/vagrant/repos/carto/lib/carto/renderer_js.js:120:23)
      at Object.CartoCSS (/vagrant/repos/carto/lib/carto/renderer_js.js:10:10)
      at render (/vagrant/repos/carto/lib/carto/renderer_js.js:280:12)
      at Context.<anonymous> (/vagrant/repos/carto/test/rendering_js.test.js:192:58)
      [...]
```
On Node v6, but it doesn't throw error on Node v0.10

**Comment by dgaubert** [comment]
However the following css input:

```
#layer[name~=".*Cairo*"] {
  marker-width: 14;
}
```

Grainstore throws:
```
Error: style0:1:957 Invalid code: [adm0name~=".*Cairo*"] { marker-width: 14;
    at Object.env.error (/vagrant/repos/grainstore/node_modules/carto/lib/carto/parser.js:241:55)
    at Object.flatten (/vagrant/repos/grainstore/node_modules/carto/lib/carto/tree/ruleset.js:158:21)
    at Object.toList (/vagrant/repos/grainstore/node_modules/carto/lib/carto/parser.js:252:44)
    at /vagrant/repos/grainstore/node_modules/carto/lib/carto/renderer.js:101:29
    at arrayMap (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:660:23)
    at Function.map (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:9571:14)
    at interceptor (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:16970:35)
    at thru (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:8812:14)
    at /vagrant/repos/grainstore/node_modules/lodash/lodash.js:4379:28
    at arrayReduce (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:704:21)
    at baseWrapperValue (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:4378:14)
    at LazyWrapper.lazyValue [as value] (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:1880:16)
    at baseWrapperValue (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:4376:25)
    at LodashWrapper.wrapperValue (/vagrant/repos/grainstore/node_modules/lodash/lodash.js:9067:14)
    at Renderer.render (/vagrant/repos/grainstore/node_modules/carto/lib/carto/renderer.js:104:14)
    at renderResolvedMml (/vagrant/repos/grainstore/lib/grainstore/mml_builder.js:219:25)
    at Function.end (/vagrant/repos/grainstore/node_modules/millstone/lib/millstone.js:991:16)
    at next (/vagrant/repos/grainstore/node_modules/step/lib/step.js:51:23)
    at /vagrant/repos/grainstore/node_modules/step/lib/step.js:83:14
    at check (/vagrant/repos/grainstore/node_modules/step/lib/step.js:99:9)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

Initially I wasn't worried about this issue due to the error in `RendererJS` was extracting options from CartoCSS to use them in GeoJSON renderer. I think we need to review `carto.RendererJS` when is running on Node v6.

It'd be nice to have a fallback in [Windshaft](https://github.com/CartoDB/Windshaft/blob/2.6.5/lib/windshaft/utils/cartocss_utils.js#L13) in order to avoid to break the normal flow when rendering a GeoJSON tile. It could look like:

```
module.exports.optionsFromCartoCSS = function(cartocss) {
    var shader;
    try {
        shader = new carto.RendererJS().render(cartocss);
    } catch (e) {
        return {};
    }
    ...
}
```

cc/ @rochoa

**Comment by dgaubert** [comment]
After talking with @rochoa offline, we should fix RendererJS to handle `=~` properly on Node v6

**Comment by dgaubert** [comment]
For the record. I've discovered one failing use case:

CartoCSS:
```
#layer0['mapnik::geometry_type'=1] { 
  marker-fill: red;
  marker-width: 10;
}
```

```
> if((8388607 & (1 << ctx.zoom)) && data.mapnik::geometry_type === 1){_value = '#ff0000';}
                                                
SyntaxError: Unexpected token :
    at Object._createFn (/vagrant/repos/Windshaft/node_modules/carto/lib/carto/renderer_js.js:153:12)
    at Object.parse (/vagrant/repos/Windshaft/node_modules/carto/lib/carto/renderer_js.js:245:40)
    at Object.setStyle (/vagrant/repos/Windshaft/node_modules/carto/lib/carto/renderer_js.js:120:23)
    at Object.CartoCSS (/vagrant/repos/Windshaft/node_modules/carto/lib/carto/renderer_js.js:10:10)
    at render (/vagrant/repos/Windshaft/node_modules/carto/lib/carto/renderer_js.js:280:12)
    at Object.module.exports.optionsFromCartoCSS (/vagrant/repos/Windshaft/lib/windshaft/utils/cartocss_utils.js:13:41)
    at /vagrant/repos/Windshaft/lib/windshaft/renderers/mapnik/geojson_renderer.js:20:38
...
```

Also it just happens on Node v6.

**Comment by dgaubert** [comment]
Oh, I've just found the surrounding cause:

- Node v0.10.26
```
$nvm run v0.10.26
Running node v0.10.26 (npm v2.14.16)
> var f = new Function("var x = false; if (x && x =~ '*.cairo.') { return true } else { return false}");
undefined
> f()
ReferenceError: Invalid left-hand side in assignment
    at eval (eval at <anonymous> (repl:1:9), <anonymous>:2:16)
    at repl:1:2
    at REPLServer.self.eval (repl.js:110:21)
    at Interface.<anonymous> (repl.js:239:12)
    at Interface.EventEmitter.emit (events.js:95:17)
    at Interface._onLine (readline.js:202:10)
    at Interface._line (readline.js:531:8)
    at Interface._ttyWrite (readline.js:760:14)
    at ReadStream.onkeypress (readline.js:99:10)
    at ReadStream.EventEmitter.emit (events.js:98:17)
>
```

- Node v6.10.0
```
$ nvm run v6.10.0
Running node v6.10.0 (npm v3.10.10)
> var f = new Function("var x = false; if (x && x =~ '*.cairo.') { return true } else { return false}");
ReferenceError: Invalid left-hand side in assignment
    at repl:1:9
    at sigintHandlersWrap (vm.js:22:35)
    at sigintHandlersWrap (vm.js:96:12)
    at ContextifyScript.Script.runInThisContext (vm.js:21:12)
    at REPLServer.defaultEval (repl.js:340:29)
    at bound (domain.js:280:14)
    at REPLServer.runBound [as eval] (domain.js:293:12)
    at REPLServer.<anonymous> (repl.js:538:10)
    at emitOne (events.js:101:20)
    at REPLServer.emit (events.js:188:7)
>
```

Basically, what is happening is: for latest V8's versions function's constructor evaluates proactively function's body, whereas in older versions it wasn't evaluated. At the end, the function is not built correctly because the operator `=~` in CartoCSS is not supported in Javascript. For example, when Javascript parses `if (x && x =~ '*.cairo.') { ... }` actually considers `=~` as two operators: `=` as assignment operator and `~` as [bitwise operator](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators#Bitwise_NOT) and it raises a `ReferenceError` because it's trying to assign a variable inside of a conditional statement.

[Windshaft](https://github.com/CartoDB/Windshaft/blob/2.6.5/lib/windshaft/utils/cartocss_utils.js#L13) creates a shader building javascript's objects and functions with CartoCSS as input but is not calling to any function dynamically created. Thus it has worked in Node v0.10.26 so far but in Node v6.10.0 fails.

We discovered this issue because a [test](https://github.com/CartoDB/Windshaft-cartodb/blob/master/test/acceptance/geojson-renderer.js#L141-L156). There are some surrounding questions: 

 - Should we support CartoCSS operators like `=~` for RendererJS?
 - Should we take care of it for the migration?
 - Should we just change that test and implement a fallback mechanism as is describe above?

Please, let me know your thoughts?

cc/ @rochoa  @rafatower @donflopez

**Comment by donflopez** [comment]
Ey! Nice catch @dgaubert! That operator is a regular expresion, isn't it?

We can transpile that to a proper javascript expresion as `x.search(/.*Cairo*/g) >= 0`

What do you think?

**Comment by dgaubert** [comment]
> That operator is a regular expresion, isn't it?
It is.

> We can transpile that to a proper javascript expresion as x.search(/.*Cairo*/g) >= 0

I'd be really nice. But I'm not sure I we have to address it now or later.

**Comment by dgaubert** [comment]
BTW I'm taking a look to determine how feasible to fix it is.

**Comment by dgaubert** [comment]
PR: #29

**Comment by dgaubert** [comment]
Closed by #29
